spec/abci++/abci++_basic_concepts_002_draft.md

```diff
@@ -1,51 +1,266 @@
+fills the above mentioned fields in `ResponsePrepareProposal`.
+
+### Tendermint timeouts in same-block execution
+
+The new same-block execution mode requires the Application to fully execute the
+prepared block at `PrepareProposal` time. This execution is synchronous, so
+Tendermint cannot make progress until the Application returns from `PrepareProposal`.
+This stands on Tendermint's critical path: if the Application takes a long time
+executing the block, the default value of _TimeoutPropose_ might not be sufficient
+to accommodate the long block execution time and non-proposer processes might time
+out and prevote `nil`, thus starting a further round unnecessarily.
+
+The Application is the best suited to provide a value for _TimeoutPropose_ so
+that the block execution time upon `PrepareProposal` fits well in the propose
+timeout interval.
+
+Currently, the Application can override the value of _TimeoutPropose_ via the
+`config.toml` file. In the future, `ConsensusParams` may have an extra field
+with the current _TimeoutPropose_ value so that the Application can adapt it at every height.
+
+## Determinism
+[&uparrow; Back to Outline](#outline)
+
+ABCI++ applications must implement deterministic finite-state machines to be
+securely replicated by the Tendermint consensus engine. This means block execution
+over the Consensus Connection must be strictly deterministic: given the same
+ordered set of transactions, all nodes will compute identical responses, for all
+successive `FinalizeBlock` calls. This is critical because the
+responses are included in the header of the next block, either via a Merkle root
+or directly, so all nodes must agree on exactly what they are.
+
+For this reason, it is recommended that applications not be exposed to any
+external user or process except via the ABCI connections to a consensus engine
+like Tendermint Core. The Application must only change its state based on input
+from block execution (`FinalizeBlock` calls), and not through
+any other kind of request. This is the only way to ensure all nodes see the same
+transactions and compute the same results.
+
+Some Applications may choose to execute the blocks that are about to be proposed
+(via `PrepareProposal`), or those that the Application is asked to validate
+(via `Processproposal`). However, the state changes caused by processing those
+proposed blocks must never replace the previous state until `FinalizeBlock` confirms
+the block decided.
+
+Additionally, vote extensions or the validation thereof (via `ExtendVote` or
+`VerifyVoteExtension`) must _never_ have side effects on the current state.
+They can only be used when their data is provided in a `RequestPrepareProposal` call.
+
+If there is some non-determinism in the state machine, consensus will eventually
+fail as nodes disagree over the correct values for the block header. The
+non-determinism must be fixed and the nodes restarted.
```

Comment:
Is it worth noting somewhere that one practical example of this is that a well-meaning application may wind up equivocating on future blocks, and thus appear Byzantine even though the fault was not explicitly malicious?

(This doesn't add any new information, but the example might help clarify why this matters)